File: craft/commands/wallet.py
import click

from craft.client import get, post
from craft.output import print_item, print_page_info, print_success, print_table


@click.command("balance")
def wallet_balance():
    """Get wallet balance."""
    data = get("/wallet")
    print_item(data)


@click.command("transactions")
@click.option("--page", default=1)
@click.option("--limit", default=20, help="Items per page (max 100)")
def wallet_transactions(page, limit):
    """List wallet transactions."""
    data = get("/wallet/transactions", params={"page": page, "limit": limit})
    items = data.get("data", data)
    if isinstance(items, dict):
        items = items.get("items", items.get("transactions", []))
    if isinstance(items, list) and items:
        rows = []
        for t in items:
            rows.append([
                t.get("id", ""),
                t.get("type", ""),
                t.get("amount", ""),
                t.get("description", t.get("note", "")),
                t.get("createdAt", ""),
            ])
        print_table(rows, ["ID", "Type", "Amount", "Description", "Date"])
        print_page_info(data, page, limit)
    else:
        print_item(data)


@click.command("topup")
@click.option("--amount", default=None, type=float, help="Amount in THB (max 100,000)")
@click.option("--reference", default=None, help="Bank transfer reference")
@click.option("--note", default=None, help="Optional note")
def wallet_topup(amount, reference, note):
    """Submit a top-up request.

    \b
    If --amount or --reference is omitted, you will be prompted.

    \b
    Examples:
      craft wallet topup --amount 500 --reference TRF123
      craft wallet topup                    # Interactive prompts
    """
    if amount is None:
        amount = click.prompt("Amount (THB)", type=float)
    if not reference:
        reference = click.prompt("Bank transfer reference")
    body = {"amount": amount, "reference": reference}
    if note:
        body["note"] = note
    data = post("/wallet/topup", body)
    print_success("Top-up submitted (pending admin approval).")
    print_item(data)

File: cmd/wallet.go
package cmd

import (
	"fmt"
	"os"
	"strconv"

	"github.com/CraftIntertech/craft-cli/internal/client"
	"github.com/CraftIntertech/craft-cli/internal/output"
	"github.com/spf13/cobra"
)

func init() {
	// wallet balance
	walletBalanceCmd := &cobra.Command{
		Use:   "balance",
		Short: "Show wallet balance",
		Run:   walletBalanceRun,
	}

	// wallet transactions
	walletTransactionsCmd := &cobra.Command{
		Use:   "transactions",
		Short: "List wallet transactions",
		Run:   walletTransactionsRun,
	}
	walletTransactionsCmd.Flags().Int("page", 1, "Page number")
	walletTransactionsCmd.Flags().Int("limit", 20, "Items per page")

	// wallet topup
	walletTopupCmd := &cobra.Command{
		Use:   "topup",
		Short: "Top up wallet balance",
		Run:   walletTopupRun,
	}
	walletTopupCmd.Flags().Float64("amount", 0, "Top-up amount")
	walletTopupCmd.Flags().String("reference", "", "Payment reference")
	walletTopupCmd.Flags().String("note", "", "Optional note")

	WalletCmd.AddCommand(walletBalanceCmd)
	WalletCmd.AddCommand(walletTransactionsCmd)
	WalletCmd.AddCommand(walletTopupCmd)
}

func walletBalanceRun(cmd *cobra.Command, args []string) {
	result := client.Get("/wallet", nil, true)
	output.PrintItem(result)
}

func walletTransactionsRun(cmd *cobra.Command, args []string) {
	page, _ := cmd.Flags().GetInt("page")
	limit, _ := cmd.Flags().GetInt("limit")

	params := map[string]string{
		"page":  strconv.Itoa(page),
		"limit": strconv.Itoa(limit),
	}

	result := client.Get("/wallet/transactions", params, true)

	data, ok := result["data"].([]interface{})
	if !ok {
		fmt.Fprintln(os.Stderr, "No transactions found.")
		return
	}

	var rows [][]string
	for _, item := range data {
		tx, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		rows = append(rows, []string{
			fmt.Sprintf("%v", tx["id"]),
			fmt.Sprintf("%v", tx["type"]),
			fmt.Sprintf("%v", tx["amount"]),
			fmt.Sprintf("%v", tx["description"]),
			fmt.Sprintf("%v", tx["date"]),
		})
	}

	output.PrintTable(rows, []string{"ID", "Type", "Amount", "Description", "Date"})

	output.PrintPageInfo(result, page, limit)
}

func walletTopupRun(cmd *cobra.Command, args []string) {
	amount, _ := cmd.Flags().GetFloat64("amount")
	reference, _ := cmd.Flags().GetString("reference")
	note, _ := cmd.Flags().GetString("note")

	if amount <= 0 {
		amountStr := PromptText("Enter top-up amount:")
		parsed, err := strconv.ParseFloat(amountStr, 64)
		if err != nil || parsed <= 0 {
			fmt.Fprintln(os.Stderr, "Invalid amount.")
			os.Exit(1)
		}
		amount = parsed
	}

	if reference == "" {
		reference = PromptText("Enter payment reference:")
	}

	body := map[string]interface{}{
		"amount":    amount,
		"reference": reference,
	}
	if note != "" {
		body["note"] = note
	}

	result := client.Post("/wallet/topup", body, true, 0)

	output.PrintSuccess("Top-up request submitted successfully.")
	output.PrintItem(result)
}
